### PR TITLE
Fixes NodePort in ipv6 with proxy-mode=ipvs

### LINKED ISSUE #68437: In ipv6-only cluster with proxy-mode=ipvs ipv4 entries are used for NodePort
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/area ipv6
/area ipvs
/sig network

**What happened**:

In an ipv6-only cluster with proxy-mode=ipvs the entries for NodePort in ipvs are created for ipv4 addresses. The target's (if there are any) consists of ipv6 addresses (wrongly) interpreted as ipv4;

```
# kubectl get svc
NAME         TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)          AGE
cgen         NodePort    fd00:4000::c698   1000::2       5001:30669/TCP   42m
...
# ipvsadm -L -n
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  127.0.0.1:30669 rr
  -> 253.0.48.0:5001              Masq    1      0          0         
TCP  192.168.0.4:30669 rr
  -> 253.0.48.0:5001              Masq    1      0          0         
TCP  [fd00:4000::c698]:5001 rr
  -> [fd00:3000:1::2]:5001        Masq    1      0          0         
  -> [fd00:3000:2::2]:5001  
...
```

Attempts to use the NodePort with ipv6 addresses are black-holed.

**What you expected to happen**:

Ipv6 addresses to the nodes should be used for NotePort and the LB-targets shall be the ipv6 address of the pods. Traffic to NodePorts shall be distributed among the (ipv6) endpoints.

**How to reproduce it (as minimally and precisely as possible)**:

* Setup an ipv6-only cluster
* Create a service with type `NodePort` with endpoints
* Try to connect to the NodePort
* Do `ipvsadm -L -n`

**Anything else we need to know?**:

The nodes have both ipv4 and ipv6 addresses and it is possible that if the nodes only had ipv6 addresses it would work (I havn't tested). But even in an ipv6-only k8s cluster the nodes almost always need ipv4 addresses for other purposes.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.0-alpha.0.1136+edd4fc25330c7f", GitCommit:"edd4fc25330c7fba7f0fd8a127b872c8614c7363", GitTreeState:"clean", BuildDate:"2018-09-07T16:28:26Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.0-alpha.0.1136+edd4fc25330c7f", GitCommit:"edd4fc25330c7fba7f0fd8a127b872c8614c7363", GitTreeState:"clean", BuildDate:"2018-09-07T16:27:51Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
  Kvm VMs on Ubuntu 18.04.01 LTS
- OS (e.g. from /etc/os-release):
  Own BusyBox based system
- Kernel (e.g. `uname -a`):
```
# uname -a
Linux vm-004 4.18.5 #1 SMP Fri Sep 7 14:40:23 CEST 2018 x86_64 GNU/Linux
```
- Install tools:
  None
- Others:

### PR DESCRIPTION
**Pull request by uablrek**
Use ipv6 addresses for NodePort with proxy-mode=ipvs in a ipv6-only cluster.

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This affects ipv6-only cluster with proxy-mode=ipvs only.

In an ipv6-only cluster the address scanning function used for setting up NodePort  rules in ipvs ipvs return ipv4 addresses.

This PR ensures that ipv6 addresses are returned by the scanning function in an ipv6-only cluster. This makes NodePort work in an ipv6-only cluster with proxy-mode=ipvs. 

**Which issue(s) this PR fixes**

Fixes #68437

**Special notes for your reviewer**:

I added a `ipv6` flag in the `netlinkHandle` struct rather than return both ipv4 and ipv6 addresses. The main reason for this is that IHO it reduces the risk for breaking the existing ipv4 function.

I am unsure how this will affect future dual-stack work, but the again I guess nobody knows and maybe it even simplifies.

Only manual testing that NodePort works in ipv6-only/ipvs is performed.

**Does this PR introduce a user-facing change?**:

NONE

```release-note
```

### CODE REVIEW

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -382,14 +382,14 @@ func NewProxier(ipt utiliptables.Interface,
...
-		ipGetter:              &realIPGetter{nl: NewNetLinkHandle()},
+		ipGetter:              &realIPGetter{nl: NewNetLinkHandle(nodeIP.To4() == nil)},
 		iptablesData:          bytes.NewBuffer(nil),
 		filterChainsData:      bytes.NewBuffer(nil),
 		natChains:             bytes.NewBuffer(nil),
 		natRules:              bytes.NewBuffer(nil),
 		filterChains:          bytes.NewBuffer(nil),
 		filterRules:           bytes.NewBuffer(nil),
-		netlinkHandle:         NewNetLinkHandle(),
+		netlinkHandle:         NewNetLinkHandle(nodeIP.To4() == nil),
```
Comment (Lion-Wei) [inline pkg/proxy/ipvs/proxier.go]: Same as above.

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -382,14 +382,14 @@ func NewProxier(ipt utiliptables.Interface,
 		healthzServer:         healthzServer,
 		ipvs:                  ipvs,
 		ipvsScheduler:         scheduler,
-		ipGetter:              &realIPGetter{nl: NewNetLinkHandle()},
+		ipGetter:              &realIPGetter{nl: NewNetLinkHandle(nodeIP.To4() == nil)},
```
Comment (Lion-Wei) [inline pkg/proxy/ipvs/proxier.go]: You can use the var `isIPv6` instead of nodeIP.To4() == nil

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (Lion-Wei) [inline pkg/proxy/ipvs/proxier.go]: Same as above

**Comment by uablrek** [comment]
@Lion-Wei God point. I will update

**Comment by uablrek** [comment]
I will also rename the parameter to NewNetLinkHandle ipv6 -> isIPv6 to have consitent naming

**Comment by Lion-Wei** [comment]
xref: #71202

**Comment by m1093782566** [comment]
I think this bug fix should be in v1.13.

/milestone v1.13

**Comment by m1093782566** [comment]
Adding label
/priority critical-urgent

as it may break an IPV6 cluster.

**Comment by AishSundar** [comment]
@uablrek  @m1093782566 was this PR meant for 1.13? we no longer accept merges into 1.13 branch and only extremely critical urgent fixes can to be CP'ed if needed.

**Comment by nikopen** [comment]
It's a pretty rare cluster conf combination for people to use directly 1.13.0 in production..

@uablrek This can be cherry picked back to 1.13.1 after it merges on master right?

It can be in the release notes that the patch will go in on 1.13.1 @marpaia 

/milestone v1.14

**Comment by uablrek** [comment]
@nikopen Not much have happened in the code as far as I know, and the base was 1.13... so I am pretty sure it can be cherry-picked.

**Comment by AishSundar** [comment]
Thanks all

/hold cancel

**Pull request merged**